### PR TITLE
KAFKA-16223: Migrate KafkaConfigBackingStoreTest to Mockito

### PR DESCRIPTION
**Pull request by johnnychhsu**
## Context
Replace Powermock with Mockito for KafkaConfigBackingStoreTest
Jira: https://issues.apache.org/jira/browse/KAFKA-16223

## Test
run `./gradlew clean :connect:runtime:test --tests KafkaConfigBackingStoreTest` and passed

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by FrankYang0529** [comment]
We may also need to remove this part.

https://github.com/apache/kafka/blob/a68a1cce824a8346509d5194e0e43a3cb36ba09a/build.gradle#L472-L477

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreTest.java [hunk 1]
```diff
@@ -194,121 +234,1293 @@ public void setUp() {
...
+//
+//        // As soon as root is rewritten, we should see a callback notifying us that we reconfigured some tasks
+//        verify(configUpdateListener).onTaskConfigUpdate(Arrays.asList(TASK_IDS.get(0), TASK_IDS.get(1)));
+
+        configStorage.stop();
+        verify(configLog).stop();
     }
 
     @Test
     public void testPutTaskConfigs() throws Exception {
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreTest.java:1468]: Please take a look at following example:

```java
    @Test
    public void testPutTaskConfigs() throws Exception {

        // Records to be read by consumer as it reads to the end of the log
        LinkedHashMap<String, byte[]> serializedConfigs = new LinkedHashMap<>();
        serializedConfigs.put(TASK_CONFIG_KEYS.get(0), CONFIGS_SERIALIZED.get(0));
        serializedConfigs.put(TASK_CONFIG_KEYS.get(1), CONFIGS_SERIALIZED.get(1));
        serializedConfigs.put(COMMIT_TASKS_CONFIG_KEYS.get(0), CONFIGS_SERIALIZED.get(2));

        // Task configs should read to end, write to the log, read to end, write root, then read to end again
        doAnswer(expectReadToEnd(new LinkedHashMap<>()))
            .doAnswer(expectReadToEnd(new LinkedHashMap<>()))
            .doAnswer(expectReadToEnd(serializedConfigs))
            .when(configLog).readToEnd();

        when(configLog.partitionCount()).thenReturn(1);

        expectConvertWriteRead(TASK_CONFIG_KEYS.get(0), KafkaConfigBackingStore.TASK_CONFIGURATION_V0, CONFIGS_SERIALIZED.get(0),
            new Struct(KafkaConfigBackingStore.TASK_CONFIGURATION_V0).put("properties", SAMPLE_CONFIGS.get(0)));

        expectConvertWriteRead(TASK_CONFIG_KEYS.get(1), KafkaConfigBackingStore.TASK_CONFIGURATION_V0, CONFIGS_SERIALIZED.get(1),
            new Struct(KafkaConfigBackingStore.TASK_CONFIGURATION_V0).put("properties", SAMPLE_CONFIGS.get(1)));

        expectConvertWriteRead(COMMIT_TASKS_CONFIG_KEYS.get(0), KafkaConfigBackingStore.CONNECTOR_TASKS_COMMIT_V0, CONFIGS_SERIALIZED.get(2),
            new Struct(KafkaConfigBackingStore.CONNECTOR_TASKS_COMMIT_V0).put("tasks", 2));

        configStorage.setupAndCreateKafkaBasedLog(TOPIC, config);
        configStorage.start();
        verifyConfigure();


        // Bootstrap as if we had already added the connector, but no tasks had been added yet
        addConnector(CONNECTOR_IDS.get(0), SAMPLE_CONFIGS.get(0), Collections.emptyList());

        // Null before writing
        ClusterConfigState configState = configStorage.snapshot();
        assertEquals(-1, configState.offset());
        assertNull(configState.taskConfig(TASK_IDS.get(0)));
        assertNull(configState.taskConfig(TASK_IDS.get(1)));

        // Writing task configs should block until all the writes have been performed and the root record update
        // has completed
        List<Map<String, String>> taskConfigs = Arrays.asList(SAMPLE_CONFIGS.get(0), SAMPLE_CONFIGS.get(1));
        configStorage.putTaskConfigs("connector1", taskConfigs);

        // Validate root config by listing all connectors and tasks
        configState = configStorage.snapshot();
        assertEquals(3, configState.offset());
        String connectorName = CONNECTOR_IDS.get(0);
        assertEquals(Collections.singletonList(connectorName), new ArrayList<>(configState.connectors()));
        assertEquals(Arrays.asList(TASK_IDS.get(0), TASK_IDS.get(1)), configState.tasks(connectorName));
        assertEquals(SAMPLE_CONFIGS.get(0), configState.taskConfig(TASK_IDS.get(0)));
        assertEquals(SAMPLE_CONFIGS.get(1), configState.taskConfig(TASK_IDS.get(1)));
        assertEquals(Collections.EMPTY_SET, configState.inconsistentConnectors());

        // As soon as root is rewritten, we should see a callback notifying us that we reconfigured some tasks
        verify(configUpdateListener).onTaskConfigUpdate(Arrays.asList(TASK_IDS.get(0), TASK_IDS.get(1)));

        configStorage.stop();
        verify(configLog).stop();
    }


    private void expectConvertWriteRead(final String configKey, final Schema valueSchema, final byte[] serialized,
                                                    final Struct value) throws Exception {
        doReturn(serialized).when(converter).fromConnectData(eq(TOPIC), eq(valueSchema), eq(value));
        doReturn(producerFuture).when(configLog).sendWithReceipt(eq(configKey), eq(serialized));
        doReturn(null).when(producerFuture).get(anyLong(), any(TimeUnit.class));
        doReturn(new SchemaAndValue(null, structToMap(value))).when(converter).toConnectData(eq(TOPIC), eq(serialized));
    }

```

Comment (johnnychhsu) [reply connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreTest.java:1468]: thanks for the sharing!
I was stuck at the expectConvertWriteRead, got different stubbing errors and not sure what could be a better solution. let me check this and modify accordingly, thanks!

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreTest.java [hunk 2]
```diff
@@ -194,51 +235,1203 @@ public void setUp() {
...
+        configStorage.refresh(0, TimeUnit.SECONDS);
+        verify(configUpdateListener).onLoggingLevelUpdate(logger3, level3);
+        verify(configUpdateListener).onLoggingLevelUpdate(logger4, level4);
+
+        configStorage.stop();
+        verify(configLog).stop();
+    }
+
+    @Test
+    public void testTaskCountRecordsAndGenerations() throws Exception {
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreTest.java:1392]:
```java
    @Test
    public void testTaskCountRecordsAndGenerations() throws Exception {

        // Records to be read by consumer as it reads to the end of the log
        LinkedHashMap<String, byte[]> serializedConfigs = new LinkedHashMap<>();
        serializedConfigs.put(TASK_CONFIG_KEYS.get(0), CONFIGS_SERIALIZED.get(0));
        serializedConfigs.put(TASK_CONFIG_KEYS.get(1), CONFIGS_SERIALIZED.get(1));
        serializedConfigs.put(COMMIT_TASKS_CONFIG_KEYS.get(0), CONFIGS_SERIALIZED.get(2));

        // Task configs should read to end, write to the log, read to end, write root, then read to end again
        doAnswer(expectReadToEnd(new LinkedHashMap<>()))
            .doAnswer(expectReadToEnd(new LinkedHashMap<>()))
            .doAnswer(expectReadToEnd(serializedConfigs))
            .doAnswer(expectReadToEnd(Collections.singletonMap(CONNECTOR_TASK_COUNT_RECORD_KEYS.get(0), CONFIGS_SERIALIZED.get(3))))
            .when(configLog).readToEnd();

        expectConvertWriteRead(TASK_CONFIG_KEYS.get(0), KafkaConfigBackingStore.TASK_CONFIGURATION_V0, CONFIGS_SERIALIZED.get(0),
            new Struct(KafkaConfigBackingStore.TASK_CONFIGURATION_V0).put("properties", SAMPLE_CONFIGS.get(0)));

        expectConvertWriteRead(TASK_CONFIG_KEYS.get(1), KafkaConfigBackingStore.TASK_CONFIGURATION_V0, CONFIGS_SERIALIZED.get(1),
            new Struct(KafkaConfigBackingStore.TASK_CONFIGURATION_V0).put("properties", SAMPLE_CONFIGS.get(1)));
        // Starts with 0 tasks, after update has 2
        expectConvertWriteRead(COMMIT_TASKS_CONFIG_KEYS.get(0), KafkaConfigBackingStore.CONNECTOR_TASKS_COMMIT_V0, CONFIGS_SERIALIZED.get(2),
            new Struct(KafkaConfigBackingStore.CONNECTOR_TASKS_COMMIT_V0).put("tasks", 2));
        // Task count records are read back after writing as well
        expectConvertWriteRead(CONNECTOR_TASK_COUNT_RECORD_KEYS.get(0), KafkaConfigBackingStore.TASK_COUNT_RECORD_V0, CONFIGS_SERIALIZED.get(3),
            new Struct(KafkaConfigBackingStore.TASK_COUNT_RECORD_V0).put("task-count", 4));


        when(configLog.partitionCount()).thenReturn(1);

        configStorage.setupAndCreateKafkaBasedLog(TOPIC, config);
        configStorage.start();
        verifyConfigure();


        // Bootstrap as if we had already added the connector, but no tasks had been added yet
        addConnector(CONNECTOR_IDS.get(0), SAMPLE_CONFIGS.get(0), Collections.emptyList());

        // Before anything is written
        String connectorName = CONNECTOR_IDS.get(0);
        ClusterConfigState configState = configStorage.snapshot();
        assertFalse(configState.pendingFencing(connectorName));
        assertNull(configState.taskCountRecord(connectorName));
        assertNull(configState.taskConfigGeneration(connectorName));

        // Writing task configs should block until all the writes have been performed and the root record update
        // has completed
        List<Map<String, String>> taskConfigs = Arrays.asList(SAMPLE_CONFIGS.get(0), SAMPLE_CONFIGS.get(1));
        configStorage.putTaskConfigs("connector1", taskConfigs);

        configState = configStorage.snapshot();
        assertEquals(3, configState.offset());
        assertTrue(configState.pendingFencing(connectorName));
        assertNull(configState.taskCountRecord(connectorName));
        assertEquals(0, (long) configState.taskConfigGeneration(connectorName));

        configStorage.putTaskCountRecord(connectorName, 4);

        configState = configStorage.snapshot();
        assertEquals(4, configState.offset());
        assertFalse(configState.pendingFencing(connectorName));
        assertEquals(4, (long) configState.taskCountRecord(connectorName));
        assertEquals(0, (long) configState.taskConfigGeneration(connectorName));

        // As soon as root is rewritten, we should see a callback notifying us that we reconfigured some tasks
        verify(configUpdateListener).onTaskConfigUpdate(Arrays.asList(TASK_IDS.get(0), TASK_IDS.get(1)));

        configStorage.stop();
        verify(configLog).stop();
    }
```

File: build.gradle [hunk 3]
```diff
@@ -472,7 +472,7 @@ subprojects {
   if (JavaVersion.current().isCompatibleWith(JavaVersion.VERSION_16)) {
     testsToExclude.addAll([
       // connect tests
-      "**/KafkaConfigBackingStoreTest.*"
+      //"**/KafkaConfigBackingStoreTest.*"
```
Comment (chia7712) [inline build.gradle]: Please remove `testsToExclude` from build.gradle. After this migration we don't use it actually.

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreTest.java [hunk 4]
```diff
@@ -194,344 +235,1463 @@ public void setUp() {
...
+        assertEquals(Arrays.asList(connectorName1, connectorName2), new ArrayList<>(configState.connectors()));
+        assertEquals(Arrays.asList(TASK_IDS.get(0), TASK_IDS.get(1)), configState.tasks(connectorName1));
+        assertEquals(Collections.singletonList(TASK_IDS.get(2)), configState.tasks(connectorName2));
+        assertEquals(SAMPLE_CONFIGS.get(0), configState.taskConfig(TASK_IDS.get(0)));
+        assertEquals(SAMPLE_CONFIGS.get(1), configState.taskConfig(TASK_IDS.get(1)));
+        assertEquals(SAMPLE_CONFIGS.get(2), configState.taskConfig(TASK_IDS.get(2)));
+        assertEquals(Collections.EMPTY_SET, configState.inconsistentConnectors());
+
+        // As soon as root is rewritten, we should see a callback notifying us that we reconfigured some tasks
+        final ArgumentCaptor<Collection<ConnectorTaskId>> capturedRecord = ArgumentCaptor.captor();
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreTest.java]: I don't think this is what we want to test. We should verify `onTaskConfigUpdate` is "executed" with "expected" inputs. For example:
```java
        verify(configUpdateListener).onTaskConfigUpdate(Arrays.asList(TASK_IDS.get(0), TASK_IDS.get(1)));
        verify(configUpdateListener).onTaskConfigUpdate(Collections.singletonList(TASK_IDS.get(2)));
```

File: build.gradle (hunk 3 above)
Comment (johnnychhsu) [reply build.gradle]: sure, let me remove that, thanks for the review!

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreTest.java (hunk 4 above)
Comment (johnnychhsu) [reply connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreTest.java]: yes you are right, I got errors when I tried to test like this, and the error reason is that the invoke times is more than 2, thus I switch it to use `verify(configUpdateListener, times)` and it passed. I think there was some mocking issue before, I tried to test with the expected arguments again and it passed this time.

**Pull request merged**